### PR TITLE
auth: refactor to include backstage identity and use more explicit types

### LINKED ISSUE #1372: Refactor Auth popup window messaging to conform to Draft RFC
There's a draft RFC that describes various flows for auth via a popup window, similar to what we're doing. https://tools.ietf.org/html/draft-sakimura-oauth-wmrm-00

Even though it's only a draft, I think it makes sense to implement what it suggests. The main and possibly only things we want to do is to use the same format for the `postMessage` response, using `{type: 'authorization_response', response: {...}}` instead of the current format, and possibly start handling the `response_mode=web_message` request parameter.

### LINKED ISSUE #1395: Identity: Use identityApi in the sidebar for the profile information
Make use of the identityApi to get profile information about the signed in user.

### LINKED ISSUE #1396: Identity: Add profile to the identityApi for the backend

### LINKED ISSUE #1397: Auth: Standardize profile format in auth backend

### PR DESCRIPTION
**Pull request by Rugvip**
Long mob branch by me, @marcuseide, and @soapraj :grin: 

Fixes #1372
Fixes #1397
Fixes #1396
Fixes #1395

This introduces a `BackstageIdentity` that contains the id and internal id token of the user.

It also standardizes the required profile information in the backend, which in turn will be used to look up users in the identity service and/or catalog.

Aaaand it fixes the main login sidebar item to actually reflect the correct login state.

And it introduces a profile to the IdentityApi

And it fixes some unmounted state access

### CODE REVIEW

File: packages/core-api/src/lib/loginPopup.ts [hunk 1]
```diff
@@ -46,24 +46,25 @@ export type LoginPopupOptions = {
...
 type AuthResult =
   | {
-      type: 'auth-result';
-      payload: any;
+      type: 'authorization_response';
+      response: unknown;
     }
   | {
-      type: 'auth-result';
+      type: 'authorization_response';
```
Comment (freben) [inline packages/core-api/src/lib/loginPopup.ts:53]: Same as the above?

File: packages/core/src/layout/SignInPage/customProvider.tsx [hunk 2]
```diff
@@ -56,6 +56,9 @@ const Component: ProviderComponent = ({ onResult }) => {
   const handleResult = ({ userId, idToken }: Data) => {
     onResult({
       userId,
+      profile: {
+        email: `${userId}@example.com`,
```
Comment (freben) [inline packages/core/src/layout/SignInPage/customProvider.tsx:60]: Legit

File: plugins/auth-backend/src/lib/PassportStrategyHelper.ts [hunk 3]
```diff
@@ -26,42 +26,52 @@ import {
...
   profile: passport.Profile,
-  params: any,
+  idToken?: string,
 ): ProfileInfo => {
-  const { displayName: name } = profile;
+  const { displayName } = profile;
 
-  let email = '';
+  let email: string | undefined = undefined;
   if (profile.emails) {
```
Comment (freben) [inline plugins/auth-backend/src/lib/PassportStrategyHelper.ts:34]: Need to check length > 0? Same below

File: plugins/auth-backend/src/providers/types.ts [hunk 4]
```diff
@@ -203,34 +231,35 @@ export type AuthInfoBase = {
...
-
-export type AuthInfoPrivate = {
+export type OAuthPrivateInfo = {
   /**
    * A refresh token issued for the signed in user.
    */
   refreshToken: string;
 };
 
+// {type: 'authorization_response', response: {...}}
```
Comment (freben) [inline plugins/auth-backend/src/providers/types.ts:241]: Remove?

File: plugins/auth-backend/src/providers/types.ts [hunk 5]
```diff
@@ -242,6 +271,22 @@ export type RedirectInfo = {
   status?: number;
 };
 
+/* 
+metadata:
```
Comment (freben) [inline plugins/auth-backend/src/providers/types.ts:275]: Remove?

File: plugins/auth-backend/src/providers/types.ts [hunk 6]
```diff
@@ -203,34 +231,35 @@ export type AuthInfoBase = {
...
 /**
  * Payload sent as a post message after the auth request is complete.
  * If successful then has a valid payload with Auth information else contains an error.
  */
-export type AuthResponse =
+export type WebMessageResponse =
   | {
-      type: 'auth-result';
-      payload: AuthInfoBase | AuthInfoWithProfile;
+      type: 'authorization_response';
```
Comment (freben) [inline plugins/auth-backend/src/providers/types.ts:249]: Déjà vu

File: plugins/auth-backend/src/providers/types.ts [hunk 7]
```diff
@@ -242,6 +271,22 @@ export type RedirectInfo = {
...
+    displayName: John Doe
+    picture: https://example.com/avatars/john.doe
+
+  identities:
+    - type: google
+      email: john.doe@gmail.com
+ */
+
+// 1. Link to identity in catalog / within backstage
+// 2. Display login information to user, i.e. sidebar popup
```
Comment (freben) [inline plugins/auth-backend/src/providers/types.ts:289]: Deserves a doc comment

**Review by freben — APPROVED** [review]

**Review by soapraj — APPROVED** [review]

File: plugins/auth-backend/src/providers/types.ts (hunk 6 above)
Comment (Rugvip) [reply plugins/auth-backend/src/providers/types.ts:249]: We tried only supplying the response or error to the function that sends the response, but that got more verbose than just listing the type a few more times.

**Pull request merged**